### DISCUSSION TITLE
How would you implement a toolbar?

### DISCUSSION THREAD

**Discussion by jossmac** [adobe/react-spectrum#4584] [Q&A] [answered]
I'd like to build something like the [Radix toolbar](https://www.radix-ui.com/docs/primitives/components/toolbar). I was hoping it could be achieved with some combination of `useListState` and `useActionGroup` + `useActionGroupItem` but the navigation and selection are bound together so you end up with tabbable "islands", rather than a single toolbar that's navigable with arrow keys regardless of group selection behaviour.

**Comment by devongovett** [comment, marked as answer]
You could use FocusScope + useFocusManager for this: https://react-spectrum.adobe.com/react-aria/FocusScope.html#usefocusmanager-example

**Reply by getDanArias** [reply]
Howdy! I bumped into this discussing while researching how to implement the `ActionGroup` component present in React Spectrum but using only React Aria. The example that you link works but it still has the tab navigation in the toolbar. The behavior in React Spectrum's `ActionGroup` is that once one tabs into the toolbar and one can only navigate with the arrow keys as the tabbing again takes us to the next component group: https://react-spectrum.adobe.com/react-spectrum/ActionGroup.html#example

May you have some insight please if this is achievable with React Aria? 🙏

**Reply by devongovett** [reply]
We do have a Toolbar in React Aria Components now. Does that do what you're looking for? https://react-spectrum.adobe.com/react-aria/Toolbar.html

**Reply by getDanArias** [reply]
Yes, this is perfect. Thanks for your reply. I missed it when going over each RAC :)

**Comment by ritz078** [comment]
I implemented it recently and the most complicated was the addition of menu but I was able to implement those too with some custom code. Focus Manager was biggest help here. See https://github.com/adobe/react-spectrum/discussions/4458 for more details

**Comment by jossmac** [comment]
Thanks for that @ritz078, I haven't yet implemented a menu but it will likely come up. Should be able to use the same roving `tabIndex` technique.

Here's where I got to:

https://github.com/adobe/react-spectrum/assets/2730833/24aa3db9-ff43-4e2b-a3cf-a0af13f9f3db

**Comment by GermanJablo** [comment]
If I use the <Button> component inside the toolbar on mobile, when I press the buttons the virtual keyboard closes. What would be the way to prevent this?

From what I see, it seems to me that this behavior is due to usePress().

Should I make a custom button for the toolbar and add all the hooks to it except usePress()?

I feel like I'm complicating something that should be simpler. Isn't there some prop or something designed for this use case?

**Reply by snowystinger** [reply]
I'm not sure what you mean about the virtual keyboard? Why is one open and what does it have to do with the toolbar? Maybe an example with steps would help?

On another note, I'm working on implementing a Toolbar hook and react aria component here https://github.com/adobe/react-spectrum/pull/5080

**Reply by GermanJablo** [reply]
> I'm not sure what you mean about the virtual keyboard? Why is one open and what does it have to do with the toolbar? Maybe an example with steps would help?

The keyboard on mobile. Generally text editors have toolbars positioned above the keyboard.

![image](https://github.com/adobe/react-spectrum/assets/43938777/d129e9b3-70cc-480b-8259-a625f0ee64e0)

> On another note, I'm working on implementing a Toolbar hook and react aria component here https://github.com/adobe/react-spectrum/pull/5080

Cool! It would be great if it supported positioning on the `floatingToolbar` selection and positioning on the mobile keyboard.
I've worked on it so I can share some hooks with you.

**Reply by snowystinger** [reply]
That's an interesting use case for the toolbar above the keyboard in mobile. It'd have to never receive focus, because that would cause the keyboard to hide (out of our control https://discussions.apple.com/thread/2739462?answerId=19353850022#19353850022).

I'll bring it up with the team and see if this is something we'd want to support

**Reply by GermanJablo** [reply]
hey @snowystinger, in the meantime, is there a way I can prevent `onPress` from closing the mobile keyboard?

right now I'm maintaining a react-aria-components button and a plain <button> with onClick... and it gets a little cumbersome.

**Reply by snowystinger** [reply]
Not currently, I tried out a few things with VoiceOver and our NumberField implementation. Moving the VO cursor to the Next/Prev buttons closes the keyboard as well.

Do you have an example without `usePress` where I could try out the behavior you're asking for?

So sorry for the delay, this slipped through the cracks. I've added it to our next agenda.

**Reply by GermanJablo** [reply]
> Do you have an example without usePress where I could try out the behavior you're asking for?

if I use a simple html button (not the react-aria button) the keyboard stays open. Not to you? Or maybe I'm not understanding your question.

> So sorry for the delay, this slipped through the cracks. I've added it to our next agenda.

Thanks!

**Reply by snowystinger** [reply]
Yeah, I don't see that staying open behavior on my iOS device without preventing focus on pointerdown, which is probably what I was referring to earlier (sorry, been a little while since I looked at this).

Possibly you could use our very hidden/secret prop https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/usePress.ts#L30
But we don't like to advertise it because it should only be used in very rare circumstances. I think in this case it is ok. It's a similar usecase to the 'clear button' in a combobox or searchfield where you want to keep editing. Just be mindful of user flows for more than touch. Try it with a screen reader, try it with a mouse, and try all the same editing with an attached keyboard.

**Reply by GermanJablo** [reply]
> Yeah, I don't see that staying open behavior on my iOS device without preventing focus on pointerdown, which is probably what I was referring to earlier (sorry, been a little while since I looked at this).

That's weird. I just tried a simple <button> on IOS and it doesn't close the keyboard. I'm not adding anything...

> Possibly you could use our very hidden/secret prop https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/usePress.ts#L30

The problem with that is that it doesn't activate the hover and press states

**Reply by snowystinger** [reply]
Could you share the example you tried? Possibly a screen recording as well. Maybe I've misinterpreted something here.

**Reply by GermanJablo** [reply]
Oh, it seems that I was calling some function inside `onClick` that caused that behavior, and the states thing was a problem with the css selector. `preventFocusOnPress` solves my problem perfectly. My only suggestion would be to expose it in the ts interface.

Sorry for the inconvenience and thank you very much for the help!

**Reply by snowystinger** [reply]
We are very hesitant to expose that one. It's not really the full story and it's usually paired with another possibly dangerous/exclusionary prop which makes it easy to exclude certain user groups from being able to complete a task.

We may eventually find a better way to signal that, but for now, making it a very conscious choice through (lack of) types is the best option we've come up with.
